base/src/main/java/uk/ac/ebi/atlas/solr/analytics/fullanalytics/FullAnalyticsDao.java

```diff
@@ -0,0 +1,125 @@
+
+    public String genesInAllAssayGroupsWithAverageExpressionSortedByCounts(String experimentAccession,
+                                                                           double expressionLevelThreshold) {
+        AnalyticsQueryBuilder queryBuilder = new AnalyticsQueryBuilder();
+        queryBuilder
+                .filterExperimentAccession(experimentAccession)
+                .queryExpressionLevel(expressionLevelThreshold)
+                .setJsonFacet(avgExpressionOverBioentityIdentifiersSortedBySpecificityJsonFacet);
+
+        LOGGER.debug(
+                "Searching for genes in {} (cutoff={}) across all assay groups...",
+                experimentAccession, expressionLevelThreshold);
+        return analyticsClient.fetchResponseAsString(queryBuilder.build());
+    }
+
+    public String genesInAssayGroupsWithAverageExpressionSortedByCounts(String experimentAccession,
+                                                                        Set<String> sliceAssayGroupIds,
+                                                                        Set<String> selectedAssayGroupIds,
+                                                                        double expressionLevelThreshold) {
+        AnalyticsQueryBuilder queryBuilder = new AnalyticsQueryBuilder();
+        queryBuilder
+                .filterExperimentAccession(experimentAccession)
+                .filterAssayGroupIds(sliceAssayGroupIds)
+                .queryExpressionLevel(expressionLevelThreshold)
+                .queryAssayGroupIds(selectedAssayGroupIds)
+                .setJsonFacet(avgExpressionOverBioentityIdentifiersSortedBySpecificityJsonFacet);
+
+        LOGGER.debug(
+                "Searching for genes in {} (cutoff={}) across {}/{} assay groups...",
+                experimentAccession, expressionLevelThreshold, selectedAssayGroupIds.size(), sliceAssayGroupIds.size());
+        return analyticsClient.fetchResponseAsString(queryBuilder.build());
+    }
+
+    public String genesNotInAssayGroupsWithMaxExpressionSortedByCounts(String experimentAccession,
+                                                                       Set<String> sliceAssayGroupIds,
+                                                                       Set<String> selectedAssayGroupIds,
+                                                                       Set<String> geneIds,
+                                                                       double expressionLevelThreshold) {
+        AnalyticsQueryBuilder queryBuilder = new AnalyticsQueryBuilder();
+        queryBuilder
+                .filterExperimentAccession(experimentAccession)
+                .filterAssayGroupIds(sliceAssayGroupIds)
+                .queryExpressionLevel(expressionLevelThreshold)
+                .queryNotAssayGroupIds(selectedAssayGroupIds)
+                .setJsonFacet(maxExpressionOverBioentityIdentifiersJsonFacet);
+
+        if (geneIds.size() < TERMS_QUERY_SIZE_THRESHOLD) {
```

Comment:
This looks dangerous. If there are too many genes, do we filter them out from results afterwards? Maybe a comment should say where that place is?

Reply:
So, if the number of genese is very big there’s a performance hit (I don’t know if it’s the network or Solr figuring out the query), so getting all gene IDs is better. In a way, `geneIds` is helping Solr to filter results so we don’t need to. I’ll add a comment to document this.